### DISCUSSION TITLE
Buttons for switching pages do not work

### DISCUSSION THREAD

**Discussion by al-zakaria** [AvaloniaUI/Avalonia#14983] [Q&A]
I have a test project to study Avalonia. And I have a question, my buttons don't work, they just don't work, they don't even let you click on them, they are just grayed out and that's it. how do I fix it? If anything, this error is in my HomePageView file.

here's a link to the project

https://github.com/Zakarayaev/AvaloniaApplication6

**Comment by IanRawley** [comment]
Your link 404's, is the repository private?

Without more information it's hard to give any real suggestions.  Being grayed out suggests the buttons are disabled, which, if they are bound to commands, suggests something is setting CanExecute to false.

**Reply by al-zakaria** [reply]
ok it really was a private repository and I fixed it. 
About additional information. I'll tell you right away that this is a test project for learning Avalonia. In general, I have a Services folder with the router files in it, and everything else is standard. The MainPageView file is the main file and there was a button to switch to the HomePage and everything worked, but when I made the TestPage and wanted to switch from HomePage to TestPage (the plan is to first go MainPage then switch to HomePage and then switch to TestPage) in the MainPageViewModel where the logic of switching was implemented an error appeared. Here it is: Error CS7036 There is no argument given that corresponds to the required parameter 'navigationService' of 'HomePageViewModel.HomePageViewModel(INavigationService)'. This is the first error. then I solved it and I don't know how correct it was but my solution was to add another constructor to the HomePageViewModel class in it and that solved the problem and switching from MainPage to HomePage worked (now I removed this constructor to see if it was the right solution I apologize if it was too brazenly of me). But in HomePageView itself the buttons were not even clicked and now the whole project has stopped working altogether, it gives 2 errors like
Error MSB3027 Could not copy "obj\Debug\net8.0/AvaloniaApplication6.dll" to "bin\Debug\net8.0/AvaloniaApplication6.dll". Exceeded retry count of 10. Failed. The file is locked by: ".NET Host (7092), .NET Host (17136), .NET Host (8992)".		

and

Error MSB3021 Unable to copy file "obj\Debug\net8.0/AvaloniaApplication6.dll" to "bin\Debug\net8.0/AvaloniaApplication6.dll". The process cannot access the file 'bin\Debug\net8.0\AvaloniaApplication6.dll' because it is being used by another process.

**Reply by stevemonaco** [reply]
These errors are a problem with your development environment locking files and not releasing. They typically persist even after closing all files in IDE and restarting the IDE. In that instance, you need to restart your computer or use something like Process Explorer from SysInternals, find what app has the lock, and then terminate it. It's a relatively rare occurrence.

**Reply by al-zakaria** [reply]
Thanks, now the project started to start at least, but what to do with the buttons?

**Reply by stevemonaco** [reply]
The buttons aren't the problem. As-is, your app doesn't build.

```
There is no argument given that corresponds to the required parameter 'navigationService' of 'HomePageViewModel.HomePageViewModel(INavigationService)'. This is the first error. then I solved it and I don't know how correct it was but my solution was to add another constructor to the HomePageViewModel class in it and that solved the problem and switching from MainPage to HomePage worked
```

The parameterless constructor is a bad idea. Your NavigationService (the specific instance created in App.axaml.cs) needs to be passed into *every* VM that can make a navigation. At least with your current design.

**Reply by thevortexcloud** [reply]
It sounds like you want DI.

https://docs.avaloniaui.net/docs/guides/implementation-guides/how-to-implement-dependency-injection

**Reply by al-zakaria** [reply]
But how to do it I just do not understand. and this link is not very helpful))) and yes to avoid this error I just added an empty constructor and now I'm asking you how to fix it all?

**Reply by thevortexcloud** [reply]
You are doing things backwards. Make a main view that hosts your routing service, and then you pass a reference to your router around. EG take a look at how ReactiveUI does it.

https://docs.avaloniaui.net/docs/concepts/reactiveui/routing

Trying to pass it into the view via the constructor is fairly annoying to do and wont work in many instances.

**Reply by al-zakaria** [reply]
But why do I need ReactiveUI? Aren't there examples on Community Toolkit?

**Reply by thevortexcloud** [reply]
You don't. The principle is the same. I was merely showing it so you could get a vague idea on how you should structure it without worrying too much about the specific implementation details.  But if you want a generic example, there is a sample here that is mostly not dependent on any specific MVVM framework (it does use some minor ReactiveUI stuff in a few places).

https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/Routing/BasicViewLocatorSample

There are _many_ ways to implement routing. But the general concept of how you display pages and what not is the same for all of them.

**Reply by timunie** [reply]
> But why do I need ReactiveUI? Aren't there examples on Community Toolkit?

RxUI was the best when Avalonia started. The CT-MVVM-Toolkit wasn't there to that time. New samples will use the MVVM Toolkit.

**Comment by al-zakaria** [comment]
Thank you very much
